pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -202,34 +202,31 @@ func (m *ManagerImpl) Stop() error {
 }
 
 func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
+	// Associates the newly created endpoint with the corresponding resource name.
+	// Stops existing endpoint if there is any.
+	existingDevs := make(map[string]pluginapi.Device)
+
+	m.mutex.Lock()
```

Comment:
There is a `newEndpoint()` call before we `Unlock()` this `mutex`. The `newEndpoint()` call does a gRPC dial which can potentially take a long time. Can we avoid holding the lock for so long?

Reply:
Yes, I realized that too:
Here is related [comments](https://github.com/kubernetes/kubernetes/issues/55179#issuecomment-342453343) 
Do you have any suggestions?

Reply:
Can we just use mutex to protect obtaining of old endpoint and its devices and adding new endpoint, but leaves newEndpoint out of the lock scope?

Reply:
Looking at this a bit more, I realized some potential race condition with this change when we have two simultaneous endpoint creations, and also a small time window after we stop the old endpoint and before we set new endpoint. To close such races, can we modify the code like this:

existingDevs := make(map[string]pluginapi.Device)
m.mutex.Lock()
old, ok := m.endpoints[r.ResourceName]
if ok && old != nil {
   existingDevs = old.devices
}
m.mutex.Unlock()
socketPath := filepath.Join(m.socketdir, r.Endpoint)
e, err := newEndpoint(socketPath, r.ResourceName, existingDevs, m.callback)
if err != nil {
   glog.Errorf("Failed to dial device plugin with request %v: %v", r, err)
   return
}
m.mutex.Lock()
ext, ok := m.endpoints[r.ResourceName]
if ext != old {
  glog.Warningf("Some other endpoint %v is added while endpoint %v is initialized", ext, e)
  m.mutex.Unlock()
  e.stop()
  return
}
m.endpoints[r.ResourceName] = e
m.mutex.Unlock()
if old != nil {
  old.stop()
}

I don't quite like this code flow either because the logic is a bit complicated. I have been started to wonder whether we should simplify the callback interface between endpoint and manager, and make endpoint totally stateless. I.e., after we merge device plugin handler and manager, whenever endpoint receives a ListWatch rpc, it just sends the whole list of devices to device plugin manager, and manager uses that list to update its internal state stored in allDevices. My only concern with that approach is that a bad device plugin can keep sending redundant device updates that causes performance issue of manager as it needs to hold its lock to compare&update device states. My current feeling for this scenario is that we perhaps want to detect and prevent such spam in a separate mechanism outside the endpoint life cycle, and simplify the interface and implementation of endpoint.

Reply:
Oh, I tried to construct the code like you suggested too, but felt the structure a little unbearable somehow.
The idea to reconsider the responsibility of manager and endpoint sounds great. Also, I noticed that you have already removed `updateCapacityFunc` in callback, and extended device plugin checkpoint data to record registered resources in #55088, so I think the change wouldn't be complicated.
As for the protection mechanism, I would suggest we update the objects in `allDevices`  stateful, maybe with a timestamps to record last updated time, to ensure minimum time interval for update.

Reply:
In the scope of this PR, perhaps we can use the mentioned code structure? I think the logic can be simplified later when we introduce a separate component to store state in manager.

Reply:
Updated with implementation above :)